File: frontend/src/services/userService.ts
import api from './api';
import { User } from '../types/user';

interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

export const userService = {
  getAllUsers: async (page: number, size: number) => {
    const response = await api.get<ApiResponse<Page<User>>>(`/users?page=${page}&size=${size}`);
    return response.data.data;
  },

  getUserByUsername: async (username: string) => {
    const response = await api.get<ApiResponse<User>>(`/users/${username}`);
    return response.data.data;
  },

  updateUser: async (username: string, userData: Partial<User>) => {
    const response = await api.put<ApiResponse<User>>(`/users/${username}`, userData);
    return response.data.data;
  },

  updateUserRoles: async (username: string, roles: string[]) => {
    const response = await api.put<ApiResponse<User>>(`/users/${username}/roles`, { roles });
    return response.data.data;
  },

  deleteUser: async (username: string) => {
    const response = await api.delete<ApiResponse<void>>(`/users/${username}`);
    return response.data.data;
  },

  searchUsers: async (keyword: string) => {
    const response = await api.get<ApiResponse<User[]>>(`/users/search?keyword=${keyword}`);
    return response.data.data;
  },

  sendEmail: async (email: string, subject: string, text: string) => {
    const response = await api.post<ApiResponse<void>>(`/users/send-email`, {
      email,
      subject,
      text,
    });
    return response.data.data;
  },

  createUser: async (userData: { username: string; password: string; email: string; name: string; roles: string[] }) => {
    const response = await api.post<ApiResponse<User>>('/users', userData);
    return response.data.data;
  },
}; 

File: backend/src/main/java/com/example/backend/service/UserService.java
package com.example.backend.service;

import com.example.backend.dto.request.UserCreateRequest;
import com.example.backend.dto.request.UserUpdateRequest;
import com.example.backend.dto.response.UserResponse;
import com.example.backend.model.enums.UserRole;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Set;

public interface UserService {
    Page<UserResponse> getAllUsers(Pageable pageable);
    UserResponse getUserByUsername(String username);
    UserResponse updateUserRoles(String username, Set<UserRole> roles);
    UserResponse updateUser(String username, UserUpdateRequest request);
    void deleteUser(String username);
    List<UserResponse> searchUsers(String keyword);
    UserResponse createUser(UserCreateRequest request);
} 